File: runner.go
package codexcw

import (
	"bufio"
	"bytes"
	"context"
	"errors"
	"fmt"
	"io"
	"os"
	"os/exec"
	"strings"
	"sync"
	"sync/atomic"
	"time"
)

const (
	defaultExecutable  = "codex"
	defaultEventBuffer = 1024
	defaultStderrLimit = 1 << 20
	defaultScanMax     = 64 << 20
)

var runCounter atomic.Uint64

// Runner starts codex exec processes and decodes their JSONL event streams.
type Runner struct {
	executable      string
	env             []string
	eventBuffer     int
	stderrLimit     int
	scanMaxBytes    int
	defaultSandbox  SandboxMode
	defaultApproval ApprovalPolicy
	now             func() time.Time
}

// Option configures a Runner.
type Option func(*Runner)

// WithExecutable changes the codex executable path. It is the primary test seam.
func WithExecutable(path string) Option {
	return func(r *Runner) {
		if path != "" {
			r.executable = path
		}
	}
}

// WithEnv appends environment variables to every child process.
func WithEnv(env ...string) Option {
	return func(r *Runner) {
		r.env = append(r.env, env...)
	}
}

// WithEventBuffer changes the per-session event channel buffer.
func WithEventBuffer(n int) Option {
	return func(r *Runner) {
		if n > 0 {
			r.eventBuffer = n
		}
	}
}

// WithStderrLimit changes the captured stderr tail size in bytes.
func WithStderrLimit(n int) Option {
	return func(r *Runner) {
		if n > 0 {
			r.stderrLimit = n
		}
	}
}

// WithScanMaxBytes changes the maximum accepted JSONL line length.
func WithScanMaxBytes(n int) Option {
	return func(r *Runner) {
		if n > 0 {
			r.scanMaxBytes = n
		}
	}
}

// WithDefaultSandbox changes the default sandbox mode.
func WithDefaultSandbox(mode SandboxMode) Option {
	return func(r *Runner) {
		if mode != "" {
			r.defaultSandbox = mode
		}
	}
}

// WithDefaultApproval changes the default approval policy.
func WithDefaultApproval(policy ApprovalPolicy) Option {
	return func(r *Runner) {
		if policy != "" {
			r.defaultApproval = policy
		}
	}
}

// New creates a Runner with safe automation defaults.
func New(opts ...Option) *Runner {
	r := &Runner{
		executable:      defaultExecutable,
		eventBuffer:     defaultEventBuffer,
		stderrLimit:     defaultStderrLimit,
		scanMaxBytes:    defaultScanMax,
		defaultSandbox:  SandboxReadOnly,
		defaultApproval: ApprovalNever,
		now:             time.Now,
	}
	for _, opt := range opts {
		opt(r)
	}
	return r
}

// Handler receives decoded events as they stream from Codex.
type Handler interface {
	// HandleCodexEvent processes one decoded event.
	HandleCodexEvent(context.Context, Event) error
}

// HandlerFunc adapts a function to Handler.
type HandlerFunc func(context.Context, Event) error

// HandleCodexEvent implements Handler.
func (f HandlerFunc) HandleCodexEvent(ctx context.Context, event Event) error {
	return f(ctx, event)
}

type runConfig struct {
	handler Handler
}

// RunOption configures one run.
type RunOption func(*runConfig)

// WithHandler registers an event handler. A handler error cancels the process.
func WithHandler(handler Handler) RunOption {
	return func(c *runConfig) {
		c.handler = handler
	}
}

// Result summarizes a completed codex exec invocation.
type Result struct {
	// RunID is the wrapper-assigned run id.
	RunID string

	// ThreadID is the Codex thread id once known.
	ThreadID string

	// FinalMessage is the last completed agent_message text.
	FinalMessage string

	// Usage is the last turn.completed usage payload.
	Usage Usage

	// Events contains every decoded event retained by Wait.
	Events []Event

	// Stderr is the captured stderr tail.
	Stderr string

	// StartedAt is the local time when collection started.
	StartedAt time.Time

	// FinishedAt is the local time when the process finished.
	FinishedAt time.Time
}

type sessionOutcome struct {
	result *Result
	err    error
}

// Session represents one running codex exec process.
type Session struct {
	// ID is the wrapper-assigned run id.
	ID string

	events chan Event
	cancel context.CancelFunc
	done   chan sessionOutcome

	mu       sync.Mutex
	threadID string
	waited   bool
	outcome  sessionOutcome
}

// Events streams decoded events until the process exits.
func (s *Session) Events() <-chan Event {
	return s.events
}

// ThreadID returns the Codex thread id once thread.started has arrived.
func (s *Session) ThreadID() string {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.threadID
}

// Cancel stops the child process through the run context.
func (s *Session) Cancel() error {
	s.cancel()
	return nil
}

// Wait waits for the process to exit and returns the final result.
func (s *Session) Wait() (*Result, error) {
	s.mu.Lock()
	if s.waited {
		outcome := s.outcome
		s.mu.Unlock()
		return outcome.result, outcome.err
	}
	s.mu.Unlock()

	outcome := <-s.done

	s.mu.Lock()
	s.waited = true
	s.outcome = outcome
	s.mu.Unlock()

	return outcome.result, outcome.err
}

// Start launches one codex exec process and returns immediately.
func (r *Runner) Start(ctx context.Context, req Request, opts ...RunOption) (*Session, error) {
	if ctx == nil {
		ctx = context.Background()
	}

	cfg := runConfig{}
	for _, opt := range opts {
		opt(&cfg)
	}

	args, stdin, cleanup, err := r.prepare(req)
	if err != nil {
		return nil, err
	}
	if cleanup != nil {
		defer func() {
			if err != nil {
				cleanup()
			}
		}()
	}

	runCtx, cancel := context.WithCancel(ctx)
	// #nosec G204 -- launching the configured Codex executable is the wrapper boundary.
	cmd := exec.CommandContext(runCtx, r.executable, args...)
	cmd.Stdin = stdin
	cmd.Env = append(os.Environ(), append(r.env, req.Env...)...)

	stdout, err := cmd.StdoutPipe()
	if err != nil {
		cancel()
		return nil, err
	}
	stderr, err := cmd.StderrPipe()
	if err != nil {
		cancel()
		return nil, err
	}

	runID := newRunID()
	session := &Session{
		ID:     runID,
		events: make(chan Event, r.eventBuffer),
		cancel: cancel,
		done:   make(chan sessionOutcome, 1),
	}

	if err = cmd.Start(); err != nil {
		cancel()
		if cleanup != nil {
			cleanup()
		}
		return nil, err
	}

	stderrTail := newTailBuffer(r.stderrLimit)
	stderrDone := make(chan struct{})
	go func() {
		_, _ = io.Copy(stderrTail, stderr)
		close(stderrDone)
	}()

	go r.collect(runCtx, session, cmd, stdout, stderrTail, stderrDone, cleanup, cfg.handler)

	return session, nil
}

// Run starts one process, drains its event stream, and waits for completion.
func (r *Runner) Run(ctx context.Context, req Request, opts ...RunOption) (*Result, error) {
	session, err := r.Start(ctx, req, opts...)
	if err != nil {
		return nil, err
	}
	for range session.Events() {
	}
	return session.Wait()
}

func (r *Runner) collect(
	ctx context.Context,
	session *Session,
	cmd *exec.Cmd,
	stdout io.Reader,
	stderrTail *tailBuffer,
	stderrDone <-chan struct{},
	cleanup func(),
	handler Handler,
) {
	startedAt := r.now()
	var (
		events       []Event
		lastEvent    *Event
		finalMessage string
		usage        Usage
		threadID     string
		runErr       error
	)

	scanner := bufio.NewScanner(stdout)
	scanner.Buffer(make([]byte, 0, 64*1024), r.scanMaxBytes)

	line := 0
	for scanner.Scan() {
		line++
		raw := bytes.TrimSpace(scanner.Bytes())
		if len(raw) == 0 {
			continue
		}
		event, err := decodeEvent(raw, session.ID, threadID, r.now())
		if err != nil {
			runErr = &DecodeError{Line: line, Raw: append([]byte(nil), raw...), Err: err}
			session.cancel()
			break
		}
		if event.ThreadStarted != nil {
			threadID = event.ThreadStarted.ThreadID
			event.ThreadID = threadID
			session.setThreadID(threadID)
		}
		if event.ThreadID == "" {
			event.ThreadID = threadID
		}
		if event.ItemCompleted != nil && event.ItemCompleted.Item.Type == ItemAgentMessage {
			finalMessage = event.ItemCompleted.Item.Text
		}
		if event.TurnCompleted != nil {
			usage = event.TurnCompleted.Usage
		}

		eventCopy := event
		lastEvent = &eventCopy
		events = append(events, event)

		select {
		case session.events <- event:
		case <-ctx.Done():
			runErr = ctx.Err()
			session.cancel()
		}
		if runErr != nil {
			break
		}

		if handler != nil {
			if err := handler.HandleCodexEvent(ctx, event); err != nil {
				runErr = &HandlerError{Err: err}
				session.cancel()
				break
			}
		}
	}

	if err := scanner.Err(); err != nil && runErr == nil {
		runErr = &DecodeError{Line: line + 1, Err: err}
		session.cancel()
	}

	waitErr := cmd.Wait()
	<-stderrDone
	if cleanup != nil {
		cleanup()
	}

	finishedAt := r.now()
	result := &Result{
		RunID:        session.ID,
		ThreadID:     threadID,
		FinalMessage: finalMessage,
		Usage:        usage,
		Events:       events,
		Stderr:       stderrTail.String(),
		StartedAt:    startedAt,
		FinishedAt:   finishedAt,
	}

	if runErr == nil {
		runErr = classifyProcessError(ctx, waitErr, result.Stderr, lastEvent)
	}
	if runErr == nil {
		runErr = classifyCodexEvent(lastEvent)
	}

	close(session.events)
	session.done <- sessionOutcome{result: result, err: runErr}
}

func (s *Session) setThreadID(threadID string) {
	s.mu.Lock()
	s.threadID = threadID
	s.mu.Unlock()
}

func classifyProcessError(ctx context.Context, waitErr error, stderr string, lastEvent *Event) error {
	if waitErr == nil {
		return nil
	}
	if ctxErr := ctx.Err(); ctxErr != nil {
		return ctxErr
	}

	var exitErr *exec.ExitError
	if errors.As(waitErr, &exitErr) {
		return &ExitError{
			Code:      exitErr.ExitCode(),
			Stderr:    stderr,
			LastEvent: lastEvent,
			Err:       waitErr,
		}
	}
	return waitErr
}

func classifyCodexEvent(lastEvent *Event) error {
	if lastEvent == nil {
		return nil
	}
	if lastEvent.Error != nil || lastEvent.TurnFailed != nil {
		return &CodexError{Event: *lastEvent}
	}
	return nil
}

func newRunID() string {
	return fmt.Sprintf("run-%d-%d", time.Now().UnixNano(), runCounter.Add(1))
}

func (r *Runner) prepare(req Request) (_ []string, _ io.Reader, cleanup func(), err error) {
	if err := validateRequest(req); err != nil {
		return nil, nil, nil, err
	}

	schemaPath := req.OutputSchemaPath
	if len(req.OutputSchema) > 0 {
		file, err := os.CreateTemp("", "codexcw-schema-*.json")
		if err != nil {
			return nil, nil, nil, err
		}
		schemaPath = file.Name()
		cleanup = func() {
			_ = os.Remove(file.Name())
		}
		if _, err := file.Write(req.OutputSchema); err != nil {
			_ = file.Close()
			cleanup()
			return nil, nil, nil, err
		}
		if err := file.Close(); err != nil {
			cleanup()
			return nil, nil, nil, err
		}
	}

	args := []string{"exec"}
	if req.ResumeID != "" || req.ResumeLast {
		args = append(args, "resume")
	}

	args = r.appendCommonArgs(args, req, schemaPath)

	if req.ResumeLast {
		args = append(args, "--last")
	}
	if req.ResumeAll {
		args = append(args, "--all")
	}

	if req.ResumeID != "" {
		args = append(args, req.ResumeID)
	}
	args = append(args, "-")

	return args, promptReader(req), cleanup, nil
}

func validateRequest(req Request) error {
	if req.Prompt == "" && req.Stdin == nil {
		return ErrPromptRequired
	}
	if len(req.OutputSchema) > 0 && req.OutputSchemaPath != "" {
		return fmt.Errorf("%w: output schema path and inline schema are mutually exclusive", ErrInvalidRequest)
	}
	resume := req.ResumeID != "" || req.ResumeLast
	if req.ResumeID != "" && req.ResumeLast {
		return fmt.Errorf("%w: resume id and resume last are mutually exclusive", ErrInvalidRequest)
	}
	if req.ResumeAll && !resume {
		return fmt.Errorf("%w: resume all requires resume id or resume last", ErrInvalidRequest)
	}
	if resume {
		if req.Dir != "" || len(req.AddDirs) > 0 || req.Profile != "" {
			return fmt.Errorf("%w: dir, add dirs, and profile are not supported by codex exec resume", ErrInvalidRequest)
		}
	}
	return nil
}

func (r *Runner) appendCommonArgs(args []string, req Request, schemaPath string) []string {
	resume := req.ResumeID != "" || req.ResumeLast

	args = append(args, "--json")
	if !resume {
		args = append(args, "--color", "never")
	}
	if req.StrictConfig {
		args = append(args, "--strict-config")
	}
	if req.Model != "" {
		args = append(args, "-m", req.Model)
	}
	if req.Profile != "" && !resume {
		args = append(args, "-p", req.Profile)
	}
	for _, feature := range req.Enable {
		args = append(args, "--enable", feature)
	}
	for _, feature := range req.Disable {
		args = append(args, "--disable", feature)
	}
	for _, image := range req.Images {
		args = append(args, "-i", image)
	}
	if !req.RequireGitRepo {
		args = append(args, "--skip-git-repo-check")
	}
	if !req.Persistent {
		args = append(args, "--ephemeral")
	}
	if req.IgnoreUserConfig {
		args = append(args, "--ignore-user-config")
	}
	if req.IgnoreRules {
		args = append(args, "--ignore-rules")
	}
	if req.DangerouslyBypassSandbox {
		args = append(args, "--dangerously-bypass-approvals-and-sandbox")
	} else {
		sandbox := req.Sandbox
		if sandbox == "" {
			sandbox = r.defaultSandbox
		}
		if resume {
			args = append(args, "-c", fmt.Sprintf("sandbox_mode=%q", string(sandbox)))
		} else {
			args = append(args, "--sandbox", string(sandbox))
		}
		approval := req.Approval
		if approval == "" {
			approval = r.defaultApproval
		}
		args = append(args, "-c", fmt.Sprintf("approval_policy=%q", string(approval)))
	}
	if req.DangerouslyBypassHooks {
		args = append(args, "--dangerously-bypass-hook-trust")
	}
	if !resume {
		if req.Dir != "" {
			args = append(args, "-C", req.Dir)
		}
		for _, dir := range req.AddDirs {
			args = append(args, "--add-dir", dir)
		}
	}
	if schemaPath != "" {
		args = append(args, "--output-schema", schemaPath)
	}
	if req.OutputLastMessagePath != "" {
		args = append(args, "-o", req.OutputLastMessagePath)
	}
	for _, override := range req.Config {
		args = append(args, "-c", override.String())
	}
	return args
}

func promptReader(req Request) io.Reader {
	switch {
	case req.Prompt != "" && req.Stdin != nil:
		return io.MultiReader(
			strings.NewReader(req.Prompt),
			strings.NewReader("\n\n<stdin>\n"),
			req.Stdin,
			strings.NewReader("\n</stdin>\n"),
		)
	case req.Prompt != "":
		return strings.NewReader(req.Prompt)
	default:
		return req.Stdin
	}
}

type tailBuffer struct {
	mu    sync.Mutex
	limit int
	data  []byte
}

func newTailBuffer(limit int) *tailBuffer {
	return &tailBuffer{limit: limit}
}

func (b *tailBuffer) Write(p []byte) (int, error) {
	b.mu.Lock()
	defer b.mu.Unlock()

	if b.limit <= 0 {
		return len(p), nil
	}
	b.data = append(b.data, p...)
	if len(b.data) > b.limit {
		b.data = append([]byte(nil), b.data[len(b.data)-b.limit:]...)
	}
	return len(p), nil
}

func (b *tailBuffer) String() string {
	b.mu.Lock()
	defer b.mu.Unlock()
	return string(b.data)
}

File: crates/codexcw/src/runner.rs
//! The [`Runner`] that spawns agent processes and decodes their output.

use std::future::Future;
use std::pin::Pin;
use std::process::{ExitStatus, Stdio};
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::{Arc, Mutex};
use std::time::{SystemTime, UNIX_EPOCH};

use tokio::io::{AsyncReadExt, AsyncWriteExt};
use tokio::process::{Child, ChildStderr, ChildStdout, Command};
use tokio::sync::mpsc;
use tokio_stream::StreamExt;
use tokio_util::codec::{FramedRead, LinesCodec};
use tokio_util::sync::CancellationToken;

use crate::args::prepare;
use crate::claude::{prepare_claude, ClaudeDecoder};
use crate::decoder::decode_event;
use crate::error::Error;
use crate::event::{Event, EventPayload, ItemKind, Usage};
use crate::request::{ApprovalPolicy, Request, SandboxMode};
use crate::session::{Completion, RunOutcome, RunResult, Session};
use crate::tail::TailBuffer;

const DEFAULT_EVENT_BUFFER: usize = 1024;
pub(crate) const DEFAULT_STDERR_LIMIT: usize = 1 << 20;
const DEFAULT_SCAN_MAX_BYTES: usize = 64 << 20;

static RUN_COUNTER: AtomicU64 = AtomicU64::new(0);

/// The CLI wrapped by a [`Runner`].
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
pub enum Agent {
    /// `codex exec --json` (the default).
    #[default]
    Codex,
    /// `claude -p --output-format stream-json`.
    Claude,
}

impl Agent {
    /// Returns the agent's default executable name.
    pub fn as_str(&self) -> &'static str {
        match self {
            Agent::Codex => "codex",
            Agent::Claude => "claude",
        }
    }
}

impl std::fmt::Display for Agent {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(self.as_str())
    }
}

/// Turns one JSONL line into zero or more decoded events.
enum EventDecoder {
    Codex,
    Claude(ClaudeDecoder),
}

impl EventDecoder {
    fn for_agent(agent: Agent) -> Self {
        match agent {
            Agent::Codex => EventDecoder::Codex,
            Agent::Claude => EventDecoder::Claude(ClaudeDecoder::default()),
        }
    }

    fn decode(
        &mut self,
        line: &[u8],
        run_id: &str,
        thread_id: &str,
        now: std::time::SystemTime,
    ) -> Result<Vec<Event>, String> {
        match self {
            EventDecoder::Codex => decode_event(line, run_id, thread_id, now).map(|e| vec![e]),
            EventDecoder::Claude(decoder) => decoder.decode(line, run_id, thread_id, now),
        }
    }
}

/// An async callback invoked for each streamed event. Returning `Err` cancels
/// the run with a [`Error::Handler`].
pub type Handler =
    Arc<dyn Fn(Event) -> Pin<Box<dyn Future<Output = Result<(), String>> + Send>> + Send + Sync>;

/// Wraps a closure into a [`Handler`].
pub fn handler<F, Fut>(f: F) -> Handler
where
    F: Fn(Event) -> Fut + Send + Sync + 'static,
    Fut: Future<Output = Result<(), String>> + Send + 'static,
{
    Arc::new(move |event| Box::pin(f(event)))
}

/// Per-run options applied to [`Runner::run_opts`] and [`Runner::start_opts`].
#[derive(Default, Clone)]
pub struct RunOptions {
    /// Event handler invoked for every decoded event.
    pub handler: Option<Handler>,
}

impl RunOptions {
    /// Returns options with the given handler set.
    pub fn with_handler(handler: Handler) -> Self {
        RunOptions {
            handler: Some(handler),
        }
    }
}

struct RunnerInner {
    agent: Agent,
    executable: String,
    env: Vec<(String, String)>,
    event_buffer: usize,
    stderr_limit: usize,
    scan_max_bytes: usize,
    default_sandbox: SandboxMode,
    default_approval: ApprovalPolicy,
}

/// Starts agent processes with safe automation defaults.
#[derive(Clone)]
pub struct Runner {
    inner: Arc<RunnerInner>,
}

impl Default for Runner {
    fn default() -> Self {
        Runner::new()
    }
}

impl Runner {
    /// Creates a runner with safe automation defaults.
    pub fn new() -> Self {
        RunnerBuilder::default().build()
    }

    /// Returns a builder for customizing the runner.
    pub fn builder() -> RunnerBuilder {
        RunnerBuilder::default()
    }

    #[cfg(test)]
    pub(crate) fn executable(&self) -> &str {
        &self.inner.executable
    }

    #[cfg(test)]
    pub(crate) fn default_sandbox(&self) -> SandboxMode {
        self.inner.default_sandbox
    }

    #[cfg(test)]
    pub(crate) fn default_approval(&self) -> ApprovalPolicy {
        self.inner.default_approval
    }

    pub(crate) fn event_buffer(&self) -> usize {
        self.inner.event_buffer
    }

    /// Launches one agent process and returns immediately.
    pub async fn start(&self, req: Request) -> Result<Session, Error> {
        self.start_opts(req, RunOptions::default()).await
    }

    /// Launches one `codex exec` process with per-run options.
    pub async fn start_opts(&self, req: Request, opts: RunOptions) -> Result<Session, Error> {
        let prepared = match self.inner.agent {
            Agent::Claude => prepare_claude(&req)?,
            Agent::Codex => prepare(
                &req,
                self.inner.default_sandbox,
                self.inner.default_approval,
            )?,
        };

        let mut command = Command::new(&self.inner.executable);
        command.args(&prepared.args);
        if let Some(dir) = &prepared.current_dir {
            command.current_dir(dir);
        }
        command.envs(self.inner.env.iter().map(|(k, v)| (k.clone(), v.clone())));
        command.envs(req.env.iter().map(|(k, v)| (k.clone(), v.clone())));
        command
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .kill_on_drop(true);

        let mut child = command
            .spawn()
            .map_err(|err| Error::Process(err.to_string()))?;

        let stdout = child.stdout.take().expect("stdout is piped");
        let stderr = child.stderr.take().expect("stderr is piped");
        if let Some(mut stdin) = child.stdin.take() {
            let data = prepared.stdin.clone();
            tokio::spawn(async move {
                let _ = stdin.write_all(&data).await;
                let _ = stdin.shutdown().await;
            });
        }

        let run_id = new_run_id();
        let (event_tx, event_rx) = mpsc::unbounded_channel();
        let (tx, rx) = mpsc::channel(self.inner.event_buffer.max(1));
        let cancel = CancellationToken::new();
        let thread_id = Arc::new(Mutex::new(String::new()));
        let completion = Arc::new(Completion::new());

        tokio::spawn(forward_events(event_rx, tx));

        let tail = Arc::new(TailBuffer::new(self.inner.stderr_limit));
        let stderr_task = tokio::spawn(drain_stderr(stderr, tail.clone()));

        tokio::spawn(collect(CollectCtx {
            child,
            stdout,
            stderr_task,
            tail,
            event_tx,
            cancel: cancel.clone(),
            completion: completion.clone(),
            handler: opts.handler,
            run_id: run_id.clone(),
            thread_id: thread_id.clone(),
            scan_max_bytes: self.inner.scan_max_bytes,
            schema_temp: prepared.schema_temp,
            decoder: EventDecoder::for_agent(self.inner.agent),
            agent: self.inner.agent,
        }));

        Ok(Session {
            id: run_id,
            rx: Some(rx),
            thread_id,
            cancel,
            completion,
        })
    }

    /// Starts one process, drains its event stream, and waits for completion.
    pub async fn run(&self, req: Request) -> Result<RunResult, Error> {
        self.run_opts(req, RunOptions::default()).await
    }

    /// Starts one process with per-run options, drains it, and waits.
    pub async fn run_opts(&self, req: Request, opts: RunOptions) -> Result<RunResult, Error> {
        let mut session = self.start_opts(req, opts).await?;
        while session.next_event().await.is_some() {}
        session.wait().await
    }
}

/// Builds a [`Runner`].
pub struct RunnerBuilder {
    agent: Agent,
    executable: Option<String>,
    env: Vec<(String, String)>,
    event_buffer: usize,
    stderr_limit: usize,
    scan_max_bytes: usize,
    default_sandbox: SandboxMode,
    default_approval: ApprovalPolicy,
}

impl Default for RunnerBuilder {
    fn default() -> Self {
        RunnerBuilder {
            agent: Agent::Codex,
            executable: None,
            env: Vec::new(),
            event_buffer: DEFAULT_EVENT_BUFFER,
            stderr_limit: DEFAULT_STDERR_LIMIT,
            scan_max_bytes: DEFAULT_SCAN_MAX_BYTES,
            default_sandbox: SandboxMode::ReadOnly,
            default_approval: ApprovalPolicy::Never,
        }
    }
}

impl RunnerBuilder {
    /// Selects the wrapped agent CLI. The default is [`Agent::Codex`].
    pub fn agent(mut self, agent: Agent) -> Self {
        self.agent = agent;
        self
    }

    /// Overrides the agent executable path. The primary test seam.
    pub fn executable(mut self, path: impl Into<String>) -> Self {
        let path = path.into();
        if !path.is_empty() {
            self.executable = Some(path);
        }
        self
    }

    /// Appends an environment variable for every child process.
    pub fn env(mut self, key: impl Into<String>, value: impl Into<String>) -> Self {
        self.env.push((key.into(), value.into()));
        self
    }

    /// Sets the per-session event channel buffer.
    pub fn event_buffer(mut self, n: usize) -> Self {
        if n > 0 {
            self.event_buffer = n;
        }
        self
    }

    /// Sets the captured stderr tail size in bytes.
    pub fn stderr_limit(mut self, n: usize) -> Self {
        self.stderr_limit = n;
        self
    }

    /// Sets the maximum accepted JSONL line length in bytes.
    pub fn scan_max_bytes(mut self, n: usize) -> Self {
        if n > 0 {
            self.scan_max_bytes = n;
        }
        self
    }

    /// Sets the default sandbox mode.
    pub fn default_sandbox(mut self, sandbox: SandboxMode) -> Self {
        self.default_sandbox = sandbox;
        self
    }

    /// Sets the default approval policy.
    pub fn default_approval(mut self, approval: ApprovalPolicy) -> Self {
        self.default_approval = approval;
        self
    }

    /// Builds the runner.
    pub fn build(self) -> Runner {
        Runner {
            inner: Arc::new(RunnerInner {
                agent: self.agent,
                executable: self
                    .executable
                    .unwrap_or_else(|| self.agent.as_str().to_string()),
                env: self.env,
                event_buffer: self.event_buffer,
                stderr_limit: self.stderr_limit,
                scan_max_bytes: self.scan_max_bytes,
                default_sandbox: self.default_sandbox,
                default_approval: self.default_approval,
            }),
        }
    }
}

struct CollectCtx {
    child: Child,
    stdout: ChildStdout,
    stderr_task: tokio::task::JoinHandle<()>,
    tail: Arc<TailBuffer>,
    event_tx: mpsc::UnboundedSender<Event>,
    cancel: CancellationToken,
    completion: Arc<Completion>,
    handler: Option<Handler>,
    run_id: String,
    thread_id: Arc<Mutex<String>>,
    scan_max_bytes: usize,
    schema_temp: Option<tempfile::NamedTempFile>,
    decoder: EventDecoder,
    agent: Agent,
}

pub(crate) async fn drain_stderr(mut stderr: ChildStderr, tail: Arc<TailBuffer>) {
    let mut buf = [0u8; 8192];
    loop {
        match stderr.read(&mut buf).await {
            Ok(0) | Err(_) => break,
            Ok(n) => tail.write(&buf[..n]),
        }
    }
}

async fn forward_events(
    mut source: mpsc::UnboundedReceiver<Event>,
    destination: mpsc::Sender<Event>,
) {
    while let Some(event) = source.recv().await {
        if destination.send(event).await.is_err() {
            break;
        }
    }
}

async fn collect(ctx: CollectCtx) {
    let CollectCtx {
        mut child,
        stdout,
        stderr_task,
        tail,
        event_tx,
        cancel,
        completion,
        handler,
        run_id,
        thread_id,
        scan_max_bytes,
        schema_temp,
        mut decoder,
        agent,
    } = ctx;

    let started_at = SystemTime::now();
    let mut events: Vec<Event> = Vec::new();
    let mut last_event: Option<Event> = None;
    let mut final_message = String::new();
    let mut usage = Usage::default();
    let mut current_thread = String::new();
    let mut run_err: Option<Error> = None;

    let mut lines = FramedRead::new(stdout, LinesCodec::new_with_max_length(scan_max_bytes));
    let mut line_no = 0usize;

    'lines: loop {
        let next = tokio::select! {
            biased;
            _ = cancel.cancelled() => None,
            line = lines.next() => line,
        };
        let Some(line_result) = next else {
            if cancel.is_cancelled() && run_err.is_none() {
                run_err = Some(Error::Cancelled);
            }
            break;
        };

        let raw_line = match line_result {
            Ok(line) => line,
            Err(err) => {
                line_no += 1;
                run_err = Some(Error::Decode {
                    line: line_no,
                    raw: None,
                    message: err.to_string(),
                });
                cancel.cancel();
                break;
            }
        };
        line_no += 1;
        let trimmed = raw_line.trim();
        if trimmed.is_empty() {
            continue;
        }

        let decoded = match decoder.decode(
            trimmed.as_bytes(),
            &run_id,
            &current_thread,
            SystemTime::now(),
        ) {
            Ok(events) => events,
            Err(message) => {
                run_err = Some(Error::Decode {
                    line: line_no,
                    raw: Some(trimmed.as_bytes().to_vec()),
                    message,
                });
                cancel.cancel();
                break;
            }
        };

        for mut event in decoded {
            if let EventPayload::ThreadStarted { thread_id: tid } = &event.payload {
                current_thread = tid.clone();
                event.thread_id = current_thread.clone();
                *thread_id.lock().expect("thread id poisoned") = current_thread.clone();
            }
            if event.thread_id.is_empty() {
                event.thread_id = current_thread.clone();
            }
            if let EventPayload::ItemCompleted(item) = &event.payload {
                if item.kind == ItemKind::AgentMessage {
                    final_message = item.text.clone();
                }
            }
            if let EventPayload::TurnCompleted { usage: turn_usage } = &event.payload {
                usage = turn_usage.clone();
            }
            if let EventPayload::TurnFailed {
                usage: turn_usage, ..
            } = &event.payload
            {
                usage = turn_usage.clone();
            }

            last_event = Some(event.clone());
            events.push(event.clone());

            let _ = event_tx.send(event.clone());

            if let Some(handler) = &handler {
                if let Err(message) = handler(event.clone()).await {
                    run_err = Some(Error::Handler(message));
                    cancel.cancel();
                    break 'lines;
                }
            }
        }
    }

    if cancel.is_cancelled() {
        let _ = child.start_kill();
    }
    let wait_result = child.wait().await;
    let _ = stderr_task.await;
    let stderr = tail.snapshot();
    drop(schema_temp);
    let finished_at = SystemTime::now();

    let report = RunResult {
        run_id,
        thread_id: current_thread,
        final_message,
        usage,
        events,
        stderr,
        started_at,
        finished_at,
    };

    if run_err.is_none() {
        run_err = classify_agent_event(agent, last_event.as_ref());
    }
    if run_err.is_none() {
        run_err = classify_process_error(&wait_result, &report.stderr, last_event.as_ref());
    }

    completion.set(RunOutcome {
        report,
        error: run_err,
    });
}

fn classify_process_error(
    wait: &std::io::Result<ExitStatus>,
    stderr: &str,
    last_event: Option<&Event>,
) -> Option<Error> {
    match wait {
        Ok(status) if status.success() => None,
        Ok(status) => Some(Error::Exit {
            code: status.code().unwrap_or(-1),
            stderr: stderr.to_string(),
            last_event: last_event.cloned().map(Box::new),
        }),
        Err(err) => Some(Error::Process(err.to_string())),
    }
}

fn classify_agent_event(agent: Agent, last_event: Option<&Event>) -> Option<Error> {
    let event = last_event?;
    match &event.payload {
        EventPayload::Error(_) | EventPayload::TurnFailed { .. } => Some(match agent {
            Agent::Codex => Error::codex_from_event(event),
            Agent::Claude => Error::claude_from_event(event),
        }),
        _ => None,
    }
}

fn new_run_id() -> String {
    let nanos = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_nanos())
        .unwrap_or(0);
    let counter = RUN_COUNTER.fetch_add(1, Ordering::Relaxed) + 1;
    format!("run-{nanos}-{counter}")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn builder_sets_fields() {
        let runner = Runner::builder()
            .executable("codex-test")
            .env("A", "B")
            .event_buffer(2)
            .default_sandbox(SandboxMode::WorkspaceWrite)
            .default_approval(ApprovalPolicy::OnRequest)
            .build();
        assert_eq!(runner.executable(), "codex-test");
        assert_eq!(runner.event_buffer(), 2);
        assert_eq!(runner.default_sandbox(), SandboxMode::WorkspaceWrite);
        assert_eq!(runner.default_approval(), ApprovalPolicy::OnRequest);
    }

    #[test]
    fn run_id_is_unique() {
        assert_ne!(new_run_id(), new_run_id());
    }
}
